### PR TITLE
Caching

### PR DESCRIPTION
**Pull request by phamann**
This re-instates the caching logic for all endpoints.

Set cache control directives if:
- Inbound request is from a Fastly node
- Inbound request is not a PUT or POST
- We are in PROD environment

If request is performing a Save or Delete via `PUT` or `POST`  we:
- Purge the corresponding `GET` request form Fastly
- Use [`Soft Purge`](https://www.fastly.com/blog/introducing-soft-purge) to boost resilience.

/cc @adgad @wheresrhys

### CODE REVIEW

**Comment by phamann** [comment]
Todo once roll out:
- [ ] Deploy Fastly VCL changes to PROD
- [ ] Update `next-myft-client` to use `next.ft.com` endpoint
- [ ] Test its working as expected
- [ ] Observe impact on origin requests

File: server/controllers/events/delete.js [hunk 1]
```diff
@@ -4,6 +4,11 @@
 
 var Events = require('../../models/tables/events');
 var validateOwner = require('../../lib/validate-owner');
+var purgeCache = require('../../lib/purge-cache');
+
+function urlToPurge(protocol, owner, type) {
+	return protocol + '://next.ft.com/__user-prefs/events/' + owner + '/' + type + '/getSinceDate/-48h';
```
Comment (adgad) [inline server/controllers/events/delete.js]: I don't think we're using this endpoint anymore on the client. It should be 
/-168h?status=new

File: server/controllers/events/save.js [hunk 2]
```diff
@@ -4,6 +4,11 @@
 
 var Events = require('../../models/tables/events');
 var validateOwner = require('../../lib/validate-owner');
+var purgeCache = require('../../lib/purge-cache');
+
+function urlToPurge(protocol, owner, type) {
+	return protocol + '://next.ft.com/__user-prefs/events/' + owner + '/' + type + '/getSinceDate/-48h';
```
Comment (adgad) [inline server/controllers/events/save.js]: ditto

**Comment by adgad** [comment]
:+1: Comments aside

File: server/controllers/events/delete.js (hunk 1 above)
Comment (phamann) [inline server/controllers/events/delete.js]: Good point should have checked first :smile:

**Pull request merged**